README.md
# ManyTypes4Py: A Benchmark Python Dataset for Machine Learning-based Type Inference
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.4719447.svg)](https://doi.org/10.5281/zenodo.4719447)

- [Intro](#intro)
- [Download](#downloading-dataset)
- [Preparation](#dataset-preparation)
- [Citing MT4Py](#citing-the-dataset)
- [Roadmap](#roadmap)

# Intro
- It has *clean* and *complete* versions (from v0.7):
  - The clean version has 5.1K **type-checked** Python repositories and 1.2M type annotations.
  - The complete version has 5.2K Python repositories and 3.3M type annotations.
- Its source files are type-checked using mypy (clean version).
- Its projects were processed in JSON-formatted files using the [LibSA4Py](https://github.com/saltudelft/libsa4py) pipeline.
- Its source files were already split into training, validation, and test sets for training ML models.
- It is de-duplicated using [CD4Py](https://github.com/saltudelft/CD4Py).
- It contains **Visible Type Hints** (VTHs), which is a deep, recursive, and dynamic analysis of types from the import statements of source files and their dependencies.
- It is published in the Data Showcase of the **MSR'21** conference.

# Downloading dataset
The latest version of the dataset is publicly available on [zenodo](https://zenodo.org/record/5244636).

# Dataset preparation
We highly recommend downloading the latest version of the dataset as described above. If you want to manually prepare the dataset, follow below instructions.

## Requirements

* Python 3.5 or newer
* Python dependencies from `scripts/requirements.txt` installed (run `pip install -r scripts/requirements.txt`)
* Install the `libsa4py` package (run `git clone https://github.com/saltudelft/libsa4py.git && cd libsa4py && pip install .`)

## Steps

0. Clone the dataset:

    ```
    python -m repo_cloner -i ./mypy-dependents-by-stars.json -o repos
    ```
    
1. To change the state of the cloned repositories to the ManyType4Py's, run the following command on the `ManyTypes4PyDataset.spec`:
    
    ```
    ./scripts/reset_commits.sh  ./ManyTypes4PyDataset.spec repos
    ``` 

2. Generate duplicate tokens for dataset using `cd4py`

    ```
    cd4py --p repos --ot tokens --od manytypes4py_dataset_duplicates.jsonl.gz --d 1024
    ```

3. Gather duplicate files from the `cd4py` output tokens, and output as a single text file (using `collect_dupes.py`)

    ```
    python3 scripts/collect_dupes.py manytypes4py_dataset_duplicates.jsonl.gz manytypes4py_dup_files.txt
    ```

4. Create a copy dataset with duplicates removed from the duplicate files collected prior (using `process_dataset.py`)

    ```
    python3 scripts/process_dataset.py repos manytypes4py_dup_files.txt [new dataset path]
    ```

5. Split dataset into test, train and validation data (using `split_dataset.py`)

    ```
    python3 scripts/split_dataset.py [new dataset path] manytypes4py_split.csv
    ```

6. To process the Python repositories and produce JSON output files, run the `libsa4py` pipeline as follows:

    ```
    libsa4py process --p [new dataset path] --o [processed projects path] --s manytypes4py_split.csv --j [WORKERS COUNT]
    ```

    Check out the `libsa4py` [README](https://github.com/saltudelft/libsa4py#usage) for more info on its usage.
    
6. Create a tar of the full dataset & artifacts in one folder

    ```
    tar -czvf [output path] [dataset artifacts path]
    ```

# Citing the dataset
If you have used the dataset in your research work, please consider citing it.

```
@inproceedings{mt4py2021,
author = {A. M. Mir and E. Latoskinas and G. Gousios},
booktitle = {IEEE/ACM 18th International Conference on Mining Software Repositories (MSR)},
title = {ManyTypes4Py: A Benchmark Python Dataset for Machine Learning-Based Type Inference},
year = {2021},
pages = {585-589},
doi = {10.1109/MSR52588.2021.00079},
publisher = {IEEE Computer Society},
month = {May}
}
```

# Roadmap
- Gathering Python projects that depend on type-checkers other than mypy, i.e., pyre, pytype, and pyright.
- Apply type annotations from [typeshed](https://github.com/python/typeshed) to the dataset.

CHANGELOG.md
# Changelog
All notable changes to the ManyTypes4Py dataset will be documented in this file.
Each version of the dataset can be downloaded on [zenodo](https://zenodo.org/search?page=1&size=20&q=conceptrecid:4044635&all_versions&sort=-version).

## [0.7] 2021-08-24
- The dataset now contains *clean* and *complete* versions. The clean version has **type-checked** Python projects by [mypy](https://mypy.readthedocs.io/). The complete version contains all Python projects in the dataset same as v0.6.
- A list of type-checked Python source files are included in `type_checked_files.txt` (using mypy).

## [0.6] 2021-04-26
- Processed projects have now the type of variables inferred by [Pyre](https://pyre-check.org/).
- Visible type hints are added for the whole dataset. They are extracted recursively from the dataset's projects and their dependencies.

## [0.5] 2021-03-12
- Added name-based visible type hints for processed projects in the `extracted_visible_types` folder.

## [0.4] - 2021-03-01
- Added type annotations coverage for processed projects and their source code files. See the `type_annot_cove` field in JSON files.
 
## [0.3] - 2021-01-29
- Fixed the path issue in `dataset_split.csv` file for some file paths that contains non-ASCII characters.
- With the new fixes to the [LibSA4py](https://github.com/saltudelft/libsa4py) tool, more source code files were processed for included Python projects.

## [0.2] - 2021-01-14
- Using the [LibSA4py](https://github.com/saltudelft/libsa4py) tool, all the Python projects were processed in JSON-formatted files.
- Using the [CD4Py](https://github.com/saltudelft/CD4Py) tool, the list of duplicate files in the dataset is provided in `duplicate_files.txt` file.
- The dataset is split into train, validation and test sets by source code files. The list of files and their corresponding set is provided in `dataset_split.csv` file.

## [0.1] - 2020-09-22
- Added a manifest file (`ManyTypes4PyDataset.spec`) that contains the URL and hash commits of 5.4K Python projects that were cloned on Sep. 17th 2020 from GitHub.


scripts/split_dataset.py
from argparse import ArgumentParser
from sklearn.model_selection import train_test_split
import os
import pandas as pd
import sys

def list_files(directory: str) -> list:
    """
    List all .py files recursively in the given directory.
    :param directory: directory to search in
    :return: list of python code files
    """
    paths = []

    for root, dirs, files in os.walk(directory):
        paths += [os.path.join(root, fname) for fname in files if fname.endswith(".py")]

    return paths

def split_file_dataframe(df, test_ratio, valid_ratio):
    """
    Splits file dataframe into train, test, validation sets.
    Returns tuple (train_df, test_df, valid_df)

    :param: df  Dataframe to split
    :param: test_ratio  Ratio of dataframe to use for test set
    :param: valid_ratio Ratio of dataframe to use for validation set
    :return: (train_df, test_df, valid_df)
    """
    train_files, test_files = train_test_split(pd.DataFrame(df['file'].unique(), columns=['file']),
                                                test_size=test_ratio)
    train_files, valid_files = train_test_split(pd.DataFrame(df[df['file'].isin(train_files.to_numpy().flatten())]['file'].unique(),
                                                                columns=['file']), test_size=valid_ratio)

    df_train = df[df['file'].isin(train_files.to_numpy().flatten())]
    print(f"Number of files in train set: {df_train.shape[0]:,}")

    df_valid = df[df['file'].isin(valid_files.to_numpy().flatten())]
    print(f"Number of files in validation set: {df_valid.shape[0]:,}")

    df_test = df[df['file'].isin(test_files.to_numpy().flatten())]
    print(f"Number of files in test set: {df_test.shape[0]:,}")

    return df_train, df_test, df_valid

if __name__ == "__main__":
    arg_parser = ArgumentParser(description="Dataset split (train, test, validation) for dataset")
    arg_parser.add_argument('dataset', type=str, help='Dataset directory')
    arg_parser.add_argument('--od', type=str, help='CSV split output path', default='dataset_split.csv')
    arg_parser.add_argument('--test', type=float, help='Test split ratio: [0.0, 1.0)', default=0.2)
    arg_parser.add_argument('--valid', type=float, help='Validation split ratio: [0.0, 1.0)', default=0.1)

    args = arg_parser.parse_args()
    DATASET_DIR = args.dataset
    OUTPUT_PATH = args.od
    TEST_SIZE = args.test
    VALID_SIZE = args.valid

    # Create DF with all code files in target dataset
    print("Retrieving Python code files from:", DATASET_DIR)
    data_files = list_files(DATASET_DIR)
    df = pd.DataFrame(data_files, columns=['file'])

    # Split and insert corresponding type of dataset
    print("Splitting Python code files to train, test & validation sets")
    df_train, df_test, df_valid = split_file_dataframe(df, TEST_SIZE, VALID_SIZE)
    df_train.insert(0, "type", "train")
    df_test.insert(0, "type", "test")
    df_valid.insert(0, "type", "valid")

    # Write recombined output
    print("Combining train, test & validation into dataframe")
    result_df = pd.concat([df_train, df_test, df_valid])

    # https://stackoverflow.com/questions/36303919/python-3-0-open-default-encoding
    # https://stackoverflow.com/questions/1052225/convert-python-filenames-to-unicode
    print("Writing dataframe to:", OUTPUT_PATH)
    result_df.to_csv(OUTPUT_PATH, header=False, index=False, encoding=sys.getfilesystemencoding())
